File: bindings/java-jni/src/main/java/org/maplibre/nativejni/render/VulkanBorrowedTextureDescriptor.java
package org.maplibre.nativejni.render;

import java.util.Objects;

/** Mutable descriptor for Vulkan caller-owned texture render targets. */
public final class VulkanBorrowedTextureDescriptor {
  private RenderTargetExtent extent = new RenderTargetExtent();
  private VulkanContextDescriptor context = new VulkanContextDescriptor();
  private NativePointer image = NativePointer.NULL;
  private NativePointer imageView = NativePointer.NULL;
  private int format;
  private int initialLayout;
  private Integer finalLayout;

  public RenderTargetExtent extent() {
    return extent;
  }

  public VulkanBorrowedTextureDescriptor extent(RenderTargetExtent extent) {
    this.extent = Objects.requireNonNull(extent, "extent");
    return this;
  }

  public VulkanContextDescriptor context() {
    return context;
  }

  public VulkanBorrowedTextureDescriptor context(VulkanContextDescriptor context) {
    this.context = Objects.requireNonNull(context, "context");
    return this;
  }

  public NativePointer image() {
    return image;
  }

  public VulkanBorrowedTextureDescriptor image(NativePointer image) {
    this.image = Objects.requireNonNull(image, "image");
    return this;
  }

  public NativePointer imageView() {
    return imageView;
  }

  public VulkanBorrowedTextureDescriptor imageView(NativePointer imageView) {
    this.imageView = Objects.requireNonNull(imageView, "imageView");
    return this;
  }

  public int format() {
    return format;
  }

  public VulkanBorrowedTextureDescriptor format(int format) {
    this.format = format;
    return this;
  }

  public int initialLayout() {
    return initialLayout;
  }

  public VulkanBorrowedTextureDescriptor initialLayout(int initialLayout) {
    this.initialLayout = initialLayout;
    return this;
  }

  public boolean hasFinalLayout() {
    return finalLayout != null;
  }

  public Integer finalLayout() {
    return finalLayout;
  }

  public VulkanBorrowedTextureDescriptor finalLayout(int finalLayout) {
    this.finalLayout = finalLayout;
    return this;
  }

  public VulkanBorrowedTextureDescriptor clearFinalLayout() {
    finalLayout = null;
    return this;
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/render/VulkanBorrowedTextureDescriptor.kt
package org.maplibre.nativeffi.render

/** Mutable descriptor for Vulkan caller-owned texture render targets. */
public class VulkanBorrowedTextureDescriptor(
  extent: RenderTargetExtent = RenderTargetExtent(),
  context: VulkanContextDescriptor = VulkanContextDescriptor(),
  image: NativePointer = NativePointer.NULL,
  imageView: NativePointer = NativePointer.NULL,
  format: Int = 0,
  initialLayout: Int = 0,
) {
  public var extent: RenderTargetExtent = extent

  public var context: VulkanContextDescriptor = context

  public var image: NativePointer = image

  public var imageView: NativePointer = imageView

  public var format: Int = format
    set(value) {
      require(value >= 0) { "format must be non-negative" }
      field = value
    }

  public var initialLayout: Int = initialLayout
    set(value) {
      require(value >= 0) { "initialLayout must be non-negative" }
      field = value
    }

  public var finalLayout: Int? = null
    set(value) {
      value?.let { require(it >= 0) { "finalLayout must be non-negative" } }
      field = value
    }

  init {
    require(format >= 0) { "format must be non-negative" }
    require(initialLayout >= 0) { "initialLayout must be non-negative" }
  }
}
